### DISCUSSION TITLE
Annotated support of Pyright

### DISCUSSION THREAD

**Discussion by medihack** [microsoft/pyright#6528] [Q&A] [answered]
[`django-stubs`](https://github.com/typeddjango/django-stubs) has a [`WithAnnotations`](https://github.com/typeddjango/django-stubs#how-do-i-annotate-cases-where-i-called-querysetannotate) to attach further attributes or methods to an existing model (by using a TypedDict).

Unfortunately, it doesn't seem to work with Pyright. When I use it like `WithAnnotations[MyModel, Foo]`, it tells me `Too many type arguments provided for "type[_T@WithAnnotations]"; expected 1 but received 2`. If I just use it as `WithAnnotations[MyModel]`, I still can't access other attributes without a typing error.

`WithAnnotate` is defined like below:

```python
from typing import Any, Generic, Mapping, TypeVar

from django.db.models.base import Model
from typing_extensions import Annotated

from radis.reports.models import Report

# Really, we would like to use TypedDict as a bound, but it's not possible
_Annotations = TypeVar("_Annotations", covariant=True, bound=Mapping[str, Any])


class Annotations(Generic[_Annotations]):
    """Use as `Annotations[MyTypedDict]`"""

    pass


_T = TypeVar("_T", bound=Model)

WithAnnotations = Annotated[_T, Annotations[_Annotations]]
"""Alias to make it easy to annotate the model `_T` as having annotations
`_Annotations` (a `TypedDict` or `Any` if not provided).

Use as `WithAnnotations[MyModel]` or `WithAnnotations[MyModel, MyTypedDict]`.
"""
```

So it uses `Annotated` under the hood (which is part of the typing module since Python 3.5 or so). `Annotated` seems to be quite implementation-specific. Does Pyright support `Annotated` in any way?

**Comment by erictraut** [comment, marked as answer]
Pyright is a standards-based type checker, and it has no specialized knowledge of any third-party libraries. That means it doesn't understand django-specific type conventions. For issues like this, our general approach is to work with the library author and typing community to extend the type system so it can support additional features within the typing spec. A good example of this is [PEP 681](https://peps.python.org/pep-0681/), which extended the type system to support popular libraries like pydantic, attrs, and various ORMs.

If you have ideas for how the type system could be extended in a general way to support django's typing requirements, I'd be interested in having that discussion.

The `Annotated` special form, which was introduced in [PEP 593](https://peps.python.org/pep-0593/) (and included in Python 3.9) provides a way to specify additional metadata for a symbol declaration statement. This metadata is not part of the official type system, so standards-compliant type checkers cannot reason about it like they can with type expressions. Pyright "supports" `Annotated` in that it understands that the first type argument to `Annotated` is a type expression. It ignores any subsequent type arguments because these represent usage-specific metadata.

**Reply by medihack** [reply]
Thanks a lot for the clarification. That explains why `django-stubs` wants the developer to [monkeypatch](https://github.com/typeddjango/django-stubs#i-cannot-use-queryset-or-manager-with-type-annotations) Django. `django-stubs` seems to currently support this feature only for Mypy.
